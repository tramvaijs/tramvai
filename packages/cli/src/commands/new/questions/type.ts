import any from '@tinkoff/utils/array/any';

type Type = 'app' | 'child-app';

const choices = [
  {
    name: 'Standalone tramvai app',
    value: 'app',
  },
  {
    name: 'Microfrontend implemented by tramvai child-app',
    value: 'child-app',
  },
];

const typeQuestion = (answer: string) => ({
  type: 'list' as const,
  name: 'type' as const,
  message: 'Choose a type for your project',
  choices,
  when: () => !any<{ name: string; value: string }>((choice) => choice.value === answer)(choices),
  default: 'app' as const,
});

export { Type, typeQuestion };
