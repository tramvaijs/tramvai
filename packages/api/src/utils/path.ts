import path from 'node:path';

export const resolveAbsolutePathForFile = ({
  file,
  sourceDir,
  rootDir,
}: {
  file: string;
  sourceDir: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(file)) {
    return file;
  }
  if (path.isAbsolute(sourceDir)) {
    return path.join(sourceDir, file);
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, sourceDir, file);
  }
  return path.resolve(rootDir, sourceDir, file);
};

export const resolveAbsolutePathForFolder = ({
  folder,
  rootDir,
}: {
  folder: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(folder)) {
    return folder;
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, folder);
  }
  return path.resolve(rootDir, folder);
};
