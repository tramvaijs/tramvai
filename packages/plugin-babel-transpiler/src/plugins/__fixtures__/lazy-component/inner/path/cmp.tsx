import React from 'react';

const InnerCmp = () => {
  return <div>inner cmp</div>;
}

export default InnerCmp;
