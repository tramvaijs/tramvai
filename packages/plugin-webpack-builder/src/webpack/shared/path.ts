import path from 'node:path';

export const resolveAbsolutePathForEntry = ({
  entry,
  sourceDir,
  rootDir,
}: {
  entry: string;
  sourceDir: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(entry)) {
    return entry;
  }
  if (path.isAbsolute(sourceDir)) {
    return path.join(sourceDir, entry);
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, sourceDir, entry);
  }
  return path.resolve(rootDir, sourceDir, entry);
};

export const resolveAbsolutePathForOutput = ({
  output,
  rootDir,
}: {
  output: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(output)) {
    return output;
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, output);
  }
  return path.resolve(rootDir, output);
};
