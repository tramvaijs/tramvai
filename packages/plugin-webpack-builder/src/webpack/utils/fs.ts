import path from 'node:path';
import fs from 'node:fs';

export const readdirAsync = async (
  dir: string,
  prefix = '',
  filelist = [] as string[],
  filter = (file: string) => true
) => {
  const files = await fs.promises.readdir(dir).catch(() => []);

  for (const file of files) {
    const filepath = path.join(dir, file);
    const stat = await fs.promises.stat(filepath);

    if (stat.isDirectory()) {
      // eslint-disable-next-line no-param-reassign
      filelist = await readdirAsync(filepath, path.join(prefix, file), filelist, filter);
    } else if (filter(file)) {
      filelist.push(path.join(prefix, file));
    }
  }

  return filelist;
};
