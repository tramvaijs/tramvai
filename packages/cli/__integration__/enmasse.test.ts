import { join, resolve } from 'path';
import { readdirSync, existsSync } from 'fs';

import { command } from 'execa';
import type { DirectoryTree } from 'directory-tree';
import dirTree from 'directory-tree';
import rimraf from 'rimraf';

const bin = join(__dirname, '../bin/platform.js');
const examplesRoot = join(__dirname, '../../../tinkoff-examples');
const examplesList = ['module', 'package', 'react-app'];

const normalizePath = (pth: string) => pth.replace(/\\/g, '/');

const sortByPath = (items: DirectoryTree[]) => {
  return items.sort((a, b) => a.path.localeCompare(b.path));
};

const getTrees = (dir: string) => {
  const trees: { original: DirectoryTree[]; less: DirectoryTree[] } = {
    original: [],
    less: [],
  };

  // определяем, как именно собирали файлы, что бы иметь возможность вырезать хэши
  const isWebpackBuild = existsSync(join(dir, 'client', 'stats.json'));

  dirTree(dir, {}, (item) => {
    const { path, name } = item;

    const newPath = path.split('.');
    const newName = name.split('.');

    if (isWebpackBuild) {
      newPath.splice(1, 1);
      newName.splice(1, 1);
    }

    trees.original.push(item);

    const newOne = {
      ...item,
      path: normalizePath(newPath.join('.')),
      name: normalizePath(newName.join('.')),
    };

    // Might differ on different systems or runs for some reason
    // @ts-ignore
    delete newOne.size;

    trees.less.push(newOne);
  });

  return {
    original: sortByPath(trees.original),
    less: sortByPath(trees.less),
  };
};

describe('enmasse', () => {
  const currentCwd = process.cwd();
  const currentProcessEnv = process.env;

  readdirSync(examplesRoot)
    .filter((path) => examplesList.includes(path))
    .forEach((file) => {
      const app = file;

      describe(`test ${app}`, () => {
        beforeAll(() => {
          const appCwd = join(examplesRoot, app);

          process.chdir(appCwd);

          rimraf.sync(resolve(appCwd, 'dist'));

          // На MacOS chalk генерирует дополнительные символы в снапшот тестах,
          // принудительно отключаем colorized output https://github.com/chalk/supports-color#info
          process.env = { ...process.env, FORCE_COLOR: '0' };
        });

        afterAll(() => {
          process.chdir(currentCwd);
          process.env = currentProcessEnv;
        });

        it(`platform --help ${app}`, async () => {
          const result = await command(`node ${bin} --no-color --help`, { shell: true });

          expect(result.exitCode).toEqual(0);
          expect(result.stdout).toMatchSnapshot();
        });

        it(`platform build ${app}`, async () => {
          const result = await command(`node ${bin} build ${app}`, { shell: true }).catch(
            (error) => {
              console.error(error.stderr);
              console.info(error.stdout);
              throw error;
            }
          );

          const output = getTrees('dist');

          expect(result.exitCode).toEqual(0);

          expect(output.less).toMatchSnapshot();

          output.original.forEach((f) => {
            expect(f.size).toBeGreaterThan(0);
          });
        }, 160000);
      });
    });
});
