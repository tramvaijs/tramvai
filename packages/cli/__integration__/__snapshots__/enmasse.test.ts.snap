// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enmasse test module platform --help module 1`] = `
"Usage: bin-init <command> [options]

Options:
  -v, --version                                 output the version number
  -h, --help                                    display help for command

Commands:
  build|b [options] <target>                    Command to build platform entities - applications, modules, libraries
  start-prod|sp [options] <target>              Command to run in production mode
  start|s [options] <target>                    Command to run in development mode
  lint|l <target>                               Command to run lint
  task|t <name> [options]                       Run task
  analyze|a [options] <target>                  Command for analyzing application
  generate|g [target] [generator]               Command for code generation
    [target] - application name
    [generator] - action|reducer|bundle|component|page|module
  new|n [options] <name>                        Command for generating a new application
  update|u [to]                                 All tramvai packages update command
    [to] - target version <latest|prerelease|X.X.X|^X.X.X> (default: latest)
  add [options] <packageName>                   Tramvai package install command
  static|st [options] <target>                  Command to export application routes to HTML pages
  benchmark|bench [options] <command> <target>  Command to run benchmark of the @tramvai/cli

  Run tramvai <command> --help
 for detailed usage of given command.
"
`;

exports[`enmasse test module platform build module 1`] = `
[
  {
    "extension": ".js",
    "name": "module_client.js",
    "path": "dist/modules/module/prerelease/module_client.js",
    "type": "file",
  },
  {
    "extension": ".js",
    "name": "module_server.js",
    "path": "dist/modules/module/prerelease/module_server.js",
    "type": "file",
  },
  {
    "extension": ".css",
    "name": "module.css",
    "path": "dist/modules/module/prerelease/module.css",
    "type": "file",
  },
  {
    "extension": ".json",
    "name": "stats.json",
    "path": "dist/modules/module/prerelease/stats.json",
    "type": "file",
  },
]
`;

exports[`enmasse test package platform --help package 1`] = `
"Usage: bin-init <command> [options]

Options:
  -v, --version                                 output the version number
  -h, --help                                    display help for command

Commands:
  build|b [options] <target>                    Command to build platform entities - applications, modules, libraries
  start-prod|sp [options] <target>              Command to run in production mode
  start|s [options] <target>                    Command to run in development mode
  lint|l <target>                               Command to run lint
  task|t <name> [options]                       Run task
  analyze|a [options] <target>                  Command for analyzing application
  generate|g [target] [generator]               Command for code generation
    [target] - application name
    [generator] - action|reducer|bundle|component|page|module
  new|n [options] <name>                        Command for generating a new application
  update|u [to]                                 All tramvai packages update command
    [to] - target version <latest|prerelease|X.X.X|^X.X.X> (default: latest)
  add [options] <packageName>                   Tramvai package install command
  static|st [options] <target>                  Command to export application routes to HTML pages
  benchmark|bench [options] <command> <target>  Command to run benchmark of the @tramvai/cli

  Run tramvai <command> --help
 for detailed usage of given command.
"
`;

exports[`enmasse test package platform build package 1`] = `
[
  {
    "extension": ".ts",
    "name": "browser.d.ts",
    "path": "dist/browser.d.ts",
    "type": "file",
  },
  {
    "extension": ".js",
    "name": "browser.js",
    "path": "dist/browser.js",
    "type": "file",
  },
  {
    "extension": ".ts",
    "name": "index.d.ts",
    "path": "dist/index.d.ts",
    "type": "file",
  },
  {
    "extension": ".js",
    "name": "index.es.js",
    "path": "dist/index.es.js",
    "type": "file",
  },
  {
    "extension": ".js",
    "name": "index.js",
    "path": "dist/index.js",
    "type": "file",
  },
]
`;

exports[`enmasse test react-app platform --help react-app 1`] = `
"Usage: bin-init <command> [options]

Options:
  -v, --version                                 output the version number
  -h, --help                                    display help for command

Commands:
  build|b [options] <target>                    Command to build platform entities - applications, modules, libraries
  start-prod|sp [options] <target>              Command to run in production mode
  start|s [options] <target>                    Command to run in development mode
  lint|l <target>                               Command to run lint
  task|t <name> [options]                       Run task
  analyze|a [options] <target>                  Command for analyzing application
  generate|g [target] [generator]               Command for code generation
    [target] - application name
    [generator] - action|reducer|bundle|component|page|module
  new|n [options] <name>                        Command for generating a new application
  update|u [to]                                 All tramvai packages update command
    [to] - target version <latest|prerelease|X.X.X|^X.X.X> (default: latest)
  add [options] <packageName>                   Tramvai package install command
  static|st [options] <target>                  Command to export application routes to HTML pages
  benchmark|bench [options] <command> <target>  Command to run benchmark of the @tramvai/cli

  Run tramvai <command> --help
 for detailed usage of given command.
"
`;

exports[`enmasse test react-app platform build react-app 1`] = `
[
  {
    "extension": ".css",
    "name": "platform.css",
    "path": "dist/client/platform.css",
    "type": "file",
  },
  {
    "extension": ".js",
    "name": "platform.js",
    "path": "dist/client/platform.js",
    "type": "file",
  },
  {
    "extension": ".map",
    "name": "platform.js.map",
    "path": "dist/client/platform.js.map",
    "type": "file",
  },
  {
    "extension": ".js",
    "name": "react.js",
    "path": "dist/client/react.js",
    "type": "file",
  },
  {
    "extension": ".map",
    "name": "react.js.map",
    "path": "dist/client/react.js.map",
    "type": "file",
  },
  {
    "extension": ".json",
    "name": "stats",
    "path": "dist/client/stats",
    "type": "file",
  },
  {
    "extension": ".js",
    "name": "tinkoff-examples_react-app_src_bundles_first_ts.chunk.js",
    "path": "dist/client/tinkoff-examples_react-app_src_bundles_first_ts.chunk.js",
    "type": "file",
  },
  {
    "extension": ".map",
    "name": "tinkoff-examples_react-app_src_bundles_first_ts.chunk.js.map",
    "path": "dist/client/tinkoff-examples_react-app_src_bundles_first_ts.chunk.js.map",
    "type": "file",
  },
  {
    "extension": ".js",
    "name": "tinkoff-examples_react-app_src_bundles_main_ts.chunk.js",
    "path": "dist/client/tinkoff-examples_react-app_src_bundles_main_ts.chunk.js",
    "type": "file",
  },
  {
    "extension": ".map",
    "name": "tinkoff-examples_react-app_src_bundles_main_ts.chunk.js.map",
    "path": "dist/client/tinkoff-examples_react-app_src_bundles_main_ts.chunk.js.map",
    "type": "file",
  },
  {
    "extension": ".js",
    "name": "tinkoff-examples_react-app_src_bundles_second_ts.chunk.js",
    "path": "dist/client/tinkoff-examples_react-app_src_bundles_second_ts.chunk.js",
    "type": "file",
  },
  {
    "extension": ".map",
    "name": "tinkoff-examples_react-app_src_bundles_second_ts.chunk.js.map",
    "path": "dist/client/tinkoff-examples_react-app_src_bundles_second_ts.chunk.js.map",
    "type": "file",
  },
  {
    "extension": ".js",
    "name": "tinkoff-examples_react-app_src_bundles_third_ts.chunk.js",
    "path": "dist/client/tinkoff-examples_react-app_src_bundles_third_ts.chunk.js",
    "type": "file",
  },
  {
    "extension": ".map",
    "name": "tinkoff-examples_react-app_src_bundles_third_ts.chunk.js.map",
    "path": "dist/client/tinkoff-examples_react-app_src_bundles_third_ts.chunk.js.map",
    "type": "file",
  },
  {
    "extension": ".js",
    "name": "server",
    "path": "dist/server/server",
    "type": "file",
  },
  {
    "extension": ".map",
    "name": "server.map",
    "path": "dist/server/server.map",
    "type": "file",
  },
  {
    "extension": ".json",
    "name": "stats",
    "path": "dist/server/stats",
    "type": "file",
  },
]
`;
